components/renku_data_services/users/db.py

```diff
@@ -0,0 +1,196 @@
+        if not email and not requested_by.is_admin:
+            raise errors.Unauthorized(message="Non-admin users cannot list all users.")
+        return await self._get_users(email)
+
+    async def _get_users(self, email: str | None = None) -> List[UserInfo]:
+        async with self.session_maker() as session:
+            stmt = select(UserORM)
+            if email:
+                stmt = stmt.where(UserORM.email == email)
+            res = await session.execute(stmt)
+            orms = res.scalars().all()
+            return [orm.dump() for orm in orms]
+
+
+class UsersSync:
+    """Sync users from Keycloak to the database."""
+
+    def __init__(
+        self,
+        session_maker: Callable[..., AsyncSession],
+    ):
+        self.session_maker = session_maker
+
+    async def _get_user(self, id) -> UserInfo | None:
+        """Get a specific user."""
+        async with self.session_maker() as session, session.begin():
+            stmt = select(UserORM).where(UserORM.keycloak_id == id)
+            res = await session.execute(stmt)
+            orm = res.scalar_one_or_none()
+            return orm.dump() if orm else None
+
+    async def _update_or_insert_user(self, user_id: str, **kwargs):
+        """Update a user or insert it if it does not exist."""
+        async with self.session_maker() as session, session.begin():
+            res = await session.execute(select(UserORM).where(UserORM.keycloak_id == user_id))
+            existing_user = res.scalar_one_or_none()
+            kwargs.pop("keycloak_id", None)
+            kwargs.pop("id", None)
+            if not existing_user:
+                new_user = UserORM(keycloak_id=user_id, **kwargs)
+                session.add(new_user)
+            else:
+                for field_name, field_value in kwargs.items():
+                    if getattr(existing_user, field_name, None) != field_value:
+                        setattr(existing_user, field_name, field_value)
+
+    async def _remove_user(self, user_id: str):
+        """Remove a user from the database."""
+        async with self.session_maker() as session, session.begin():
+            logging.info(f"Removing user with ID {user_id}")
+            stmt = delete(UserORM).where(UserORM.keycloak_id == user_id)
+            await session.execute(stmt)
+
+    async def users_sync(self, kc_api: IKeycloakAPI):
+        """Sync all users from Keycloak into the users database."""
+        async with self.session_maker() as session, session.begin():
+            logging.info("Starting a total user database sync.")
+            kc_users = kc_api.get_users()
+            for raw_kc_user in kc_users:
+                kc_user = UserInfo.from_kc_user_payload(raw_kc_user)
+                db_user = await self._get_user(kc_user.id)
+                if db_user != kc_user:
+                    await self._update_or_insert_user(kc_user.id, **asdict(kc_user))
+
+    async def events_sync(self, kc_api: IKeycloakAPI):
```

Comment:
So partial application of the events is fine. And multiple applications of the same events is also fine. I just keep the timestamp in the db to not always have to process all events.

If KC dies halfway through then only some events would be processed and partial data would be available  in the db. I.e. if someone updated their first name and last name then only the first name would be updated but not their last. A registration is a single event in KC so a new user getting created should not get a partial update. Either we will get the user or not. I also make sure to sort the evens by time so that re-application of the same set of events results in the same final result.

I can add some tests for this. I.e. definitely test what happens when the same list of events is applied twice. And I think it should also be easy to test what happens when Keycloak crashes. Or at least I can modify the Dummy Keycloak API to be able to artificially cause this.

One more thing is that we have one single transaction for processing all events. So if a single event fails then all the other events that were processed will be undone. Since we can handle re-processing the same set of events it is probably better to have a transaction per event. I will open a follow up PR to add tests. I will look into these scenarios there.

Reply:
might be a bit of an edgecase, but if there's an update followed by a delete, and then some other operation fails, on replay you'd have an update being done to a deleted user, would this cause an issue?